programs/marinade-finance/src/instructions/user/withdraw_stake_account.rs

```diff
@@ -0,0 +1,302 @@
+
+        // require the stake is not in emergency_unstake
+        require_neq!(
+            stake.is_emergency_unstaking,
+            0,
+            MarinadeError::StakeAccountIsEmergencyUnstaking
+        );
+
+        // require stake is active (deactivation_epoch == u64::MAX)
+        let delegation = self.stake_account.delegation().ok_or_else(|| {
+            error!(MarinadeError::RequiredDelegatedStake).with_account_name("stake_account")
+        })?;
+        require_eq!(
+            delegation.deactivation_epoch,
+            std::u64::MAX,
+            MarinadeError::RequiredActiveStake
+        );
+
+        let mut validator = self
+            .state
+            .validator_system
+            .get(&self.validator_list.data.as_ref().borrow(), validator_index)?;
+
+        // check currently_staked in this account & validator vote-key
+        check_stake_amount_and_validator(
+            &self.stake_account,
+            stake.last_update_delegated_lamports,
+            &validator.validator_account,
+        )?;
+
+        // compute how many lamport the burned mSOL represents
+        require_gt!(msol_amount, self.state.min_withdraw);
+        let sol_value = self.state.calc_lamports_from_msol_amount(msol_amount)?;
+        // apply withdraw_stake_account_fee to avoid economical attacks
+        // withdraw_stake_account_fee must be >= one epoch staking rewards
+        let withdraw_stake_account_fee_lamports =
+            self.state.withdraw_stake_account_fee.apply(sol_value);
+        // the fee value will be burned but not delivered, thus increasing mSOL value slightly for all mSOL holders
+        let split_lamports = sol_value - withdraw_stake_account_fee_lamports;
+        // check withdraw amount >= WithdrawAmountIsTooLow
+        require_gte!(
+            split_lamports,
+            self.state.min_withdraw,
+            MarinadeError::WithdrawAmountIsTooLow
+        );
+        const MIN_STAKE_ACCOUNT_LAMPORTS: u64 = LAMPORTS_PER_SOL;
+        // Simplification, we always deliver a splitted account, so some lamports must remain in the original account
+        // check that after split, the amount remaining in the stake account is >= MIN_STAKE_ACCOUNT_LAMPORTS
+        require_gte!(
+            stake
+                .last_update_delegated_lamports
+                .saturating_sub(split_lamports),
+            MIN_STAKE_ACCOUNT_LAMPORTS,
+            MarinadeError::StakeAccountRemainderTooLow
+        );
+
+        // burn mSOL
+        burn(
+            CpiContext::new(
+                self.token_program.to_account_info(),
+                Burn {
+                    mint: self.msol_mint.to_account_info(),
+                    from: self.burn_msol_from.to_account_info(),
+                    authority: self.burn_msol_authority.to_account_info(),
+                },
+            ),
+            msol_amount,
+        )?;
+        self.state.on_msol_burn(msol_amount)?;
+
```

Comment:
Please add returning whole stake without splitting. For the sake of completeness

Reply:
No, that will complicate the code unnecessarily. Max withdraw-stake-account is biggest stake account - 1 SOL
Always use the split_stake_account, always return the stake there.